src/evict/evict_lru.c

```diff
@@ -1870,9 +1872,27 @@ __evict_walk_file(WT_SESSION_IMPL *session,
 
 		/* Pages that are empty or from dead trees are fast-tracked. */
 		if (__wt_page_is_empty(page) ||
-		    F_ISSET(session->dhandle, WT_DHANDLE_DEAD))
+		    F_ISSET(session->dhandle, WT_DHANDLE_DEAD) ||
+		    F_ISSET(btree, WT_BTREE_LOOKASIDE))
 			goto fast;
 
+		/*
+		 * If we're blocked by eviction (and going to consider
+		 * lookaside), and the only thing preventing a clean page from
+		 * being evicted is that it contains historical data, mark it
+		 * dirty so we can do lookaside eviction.
+		 */
+		if (F_ISSET(cache, WT_CACHE_EVICT_CLEAN_HARD |
+		    WT_CACHE_EVICT_DIRTY_HARD) &&
+		    !F_ISSET(conn, WT_CONN_EVICTION_NO_LOOKASIDE) &&
+		    F_ISSET(cache, WT_CACHE_EVICT_DIRTY) &&
+		    !modified && page->modify != NULL &&
+		    !__wt_txn_visible_all(session, page->modify->rec_max_txn,
+		    WT_TIMESTAMP_NULL(&page->modify->rec_max_timestamp))) {
+			__wt_page_only_modify_set(session, page);
+			modified = true;
+		}
+
```

Comment:
This clause is pretty obscure.

What does it mean to be "blocked by eviction"?

I think the `F_ISSET(cache, WT_CACHE_EVICT_DIRTY)` test is intended to mean "if we mark this page dirty will it be chosen?" but that test is only a few lines further down in the same function. Does it have some other meaning, or couldn't we just jump to "fast"?

Instead of calling `__wt_page_only_modify_set()` here, would it be better to enhance the final clause of `__wt_page_can_evict()` with some version of this test?

Finally, it seems like this test will potentially fail the subsequent check: `If the oldest transaction hasn't changed since the last time this page was written, it's unlikely we can make progress`. Doesn't that minimize the usefulness of this check? (If we shuffled the test order a little in this function, I think we could just jump to the "fast" label if this clause tests true.)

Reply:
I've added the `goto fast` and tried to clarify the comment.  FWIW, this change was necessary to avoid hangs in testing.

Reply:
Also note that we have to mark the page dirty: it isn't just about the eviction checks, we have to call reconciliation on the page for this to be useful.